### ISSUE TITLE
OLS autocomplete with `completion-replace = true` removes next char

### ISSUE THREAD

**Issue by roignpar** [helix-editor/helix#11933]
### Summary

When using [ols](https://github.com/DanielGavin/ols) (the Odin Language Server) with Helix, every time an autocomplete option is selected and inserted the next character is removed. If there is no next character Helix crashes. This only seems to happen when `completion-replace = true`.
I have seen the crash with other LSPs, but not the removal.

### Reproduction Steps

![Screencast From 2024-10-24 12-41-54](https://github.com/user-attachments/assets/de41a686-847a-4c58-9a50-64bb96d8b88a)
With ols installed and available to Helix, open an `.odin` file and try to insert an autocomplete suggestion.

### Helix log

The full log file is very large, I will only post the messages around the time the screencast was recorded. I tried activating `RUST_BACKTRACE`, but that didn't help much apparently.
<details><summary>~/.cache/helix/helix.log</summary>

```
2024-10-24T12:41:17.431 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:19.978 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:19.978 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:21.866 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:21.866 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:22.629 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:22.926 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:24.584 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:41:38.119 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:41:38.660 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:38.990 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:41:57.379 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:41:57.379 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:01.148 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:01.149 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:01.149 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:03.816 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:04.561 helix_lsp::transport [ERROR] ols err: <- StreamClosed
2024-10-24T12:42:04.561 helix_lsp::transport [ERROR] err: <- Other(failed to send a message to server

Caused by:
    channel closed)
2024-10-24T12:42:45.058 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:42:46.389 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:42:46.389 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:42:47.742 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:42:48.607 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:50.366 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:50.501 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:50.851 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:42:51.683 helix_lsp::transport [ERROR] ols err: <- StreamClosed
2024-10-24T12:42:51.684 helix_lsp::transport [ERROR] err: <- Other(failed to send a message to server

Caused by:
    channel closed

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>)
2024-10-24T12:43:06.437 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:43:09.421 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:43:09.421 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:43:11.226 helix_lsp::util [WARN] LSP position Position { line: 7, character: 0 } out of range assuming EOF
2024-10-24T12:43:12.131 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:43:12.945 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:43:13.689 helix_lsp::util [WARN] LSP position Position { line: 6, character: 0 } out of range assuming EOF
2024-10-24T12:43:15.082 helix_lsp::transport [ERROR] ols err: <- StreamClosed
2024-10-24T12:43:15.082 helix_lsp::transport [ERROR] err: <- Other(failed to send a message to server

Caused by:
    channel closed

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: <unknown>
  15: <unknown>)

```

</details>

### Platform

Linux

### Terminal Emulator

alacritty 0.14.0 (22a44757) + zellij 0.40.1

### Installation Method

pacman

### Helix Version

helix 24.7

**Comment by RoloEdits** [comment]
This issue might have been fixed in #11266. If you are able to build from source to try on or after that commit that would be helpful to confirm so.

**Comment by roignpar** [comment]
Indeed, the issue seems to be fixed on current master. Any clue when the next release will be?

**Comment by RoloEdits** [comment]
Hmm, actually, maybe the panic was fixed, but I think there could still be an issue here.

Using `rust-analyzer` I am also seeing the eating of characters when `completion-replace = true`. 

Can you confirm that the character removal was also fixed, not just the panic?

**Comment by roignpar** [comment]
I think it depends on what you are completing. `completion-replace = true` does exactly that, replace any word with the completion.
```rust
std::|Something
```
`|` being the cursor. If a completion is inserted at cursor `Something` will be removed.
```rust
std::|()
```
In this case `()` shouldn't be removed when inserting a completion at `|`.

Another possible bug I'm seeing now while testing again:
```rust
std::|Something
```
When inserting a completion at `|` the cursor is placed one character before the end of the inserted word:
```rust
std::syn|c // should be std::sync|
```
Just tried this again with latest master.

**Comment by RoloEdits** [comment]
Doing some testing of 

https://github.com/helix-editor/helix/blob/b53dafe326d51b7f64a1c52379e3d4150cd2991e/helix-lsp/src/lib.rs#L292-L313
and
https://github.com/helix-editor/helix/blob/b53dafe326d51b7f64a1c52379e3d4150cd2991e/helix-lsp/src/lib.rs#L276-L291

and found that the values of the calculated function(the second one) don't always match the other where the value is supposed to be passed in.

Checking if its replace mode and always calculating seemed to always produced the desired outcome(at least I haven't found any yet outliers yet):
```rust
    fn completion_range(
        text: RopeSlice,
        edit_offset: Option<(i128, i128)>,
        replace_mode: bool,
        cursor: usize,
    ) -> Option<(usize, usize)> {
        let res = if replace_mode {
            find_completion_range(text, replace_mode, cursor)
        } else {
            match edit_offset {
                Some((start_offset, end_offset)) => {
                    let start_offset = cursor as i128 + start_offset;
                    if start_offset < 0 {
                        return None;
                    }
                    let end_offset = cursor as i128 + end_offset;
                    if end_offset > text.len_chars() as i128 {
                        return None;
                    }

                    (start_offset as usize, end_offset as usize)
                }
                None => find_completion_range(text, replace_mode, cursor),
            }
        };

        Some(res)
    }
```

Im not entirely sure but it seems like the passed in values are from the lsp itself? Kind of hard to follow exactly where the `.range` is coming from as I followed this up the call stack. I did notice as well that in some of the logs from `rust-analyzer` have an error were it mentions truncation being done as there was some out of bounds happening?

So i'm not sure if this is a helix thing or a rust analyzer thing, or both. Have you noticed this behavior with odin's lsp as well? 

I build `rust-analyzer` from source so I can enable more native optimizations, but when trying the downloaded binary I also noticed it. But the one that ships with `rustup` thats dated to the 1.82 release cycle didn't appear to have the same issues. 

I think we can either open this issue again, or even make a new one, so that we can get other people looking for root of this issue?

**Comment by RoloEdits** [comment]
Here are some gif's of the behavior I am experiencing:

![helix_rust-analyzer](https://github.com/user-attachments/assets/bc3b62fe-ba69-4e7c-9214-ed0553e2f5f7)

![helix_rust-analyzer-2](https://github.com/user-attachments/assets/41d1d825-b855-4907-96bb-9340cbc86c42)

**Comment by roignpar** [comment]
It looks just like the original issue I had. My guess is it might be related to the version of rust-analyzer you are using, but maybe it is worth opening a new issue to get the attention of someone more knowledgeable.

**Comment by RoloEdits** [comment]
That is to say you no longer experience this in OLS? If so, this might be a `rust-analyzer` bug.

**Comment by roignpar** [comment]
> That is to say you no longer experience this in OLS? If so, this might be a `rust-analyzer` bug.

Yes, when using the master branch of Helix I no longer experience this with OLS.
